In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [5]:
data= pd.read_csv("C:\\Users\\Abhishek\\Downloads\\house_sales_prices.csv")

In [6]:
data.head()

LotFrontage  LotArea  YearBuilt  TotalBsmtSF  BedroomAbvGr  KitchenAbvGr  \
0         65.0     8450       2003          856             3             1   
1         80.0     9600       1976         1262             3             1   
2         68.0    11250       2001          920             3             1   
3         60.0     9550       1915          756             3             1   
4         84.0    14260       2000         1145             4             1   

   TotRmsAbvGrd Street CentralAir  GarageCars  GarageArea  YrSold SaleType  \
0             8   Pave          Y           2         548    2008       WD   
1             6   Pave          Y           2         460    2007       WD   
2             6   Pave          Y           2         608    2008       WD   
3             7   Pave          Y           3         642    2006       WD   
4             9   Pave          Y           3         836    2008       WD   

  SaleCondition  SalePrice  
0        Normal     208500  
1        Normal     181500  
2        Normal     223500  
3       Abnorml     140000  
4        Normal     250000

Finding the Null value

In [7]:
data.isnull().sum()

LotFrontage      259
LotArea            0
YearBuilt          0
TotalBsmtSF        0
BedroomAbvGr       0
KitchenAbvGr       0
TotRmsAbvGrd       0
Street             0
CentralAir         0
GarageCars         0
GarageArea         0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
dtype: int64

Dropping the Null values

In [8]:
data.dropna(inplace=True)

In [9]:
data.isnull().sum()

LotFrontage      0
LotArea          0
YearBuilt        0
TotalBsmtSF      0
BedroomAbvGr     0
KitchenAbvGr     0
TotRmsAbvGrd     0
Street           0
CentralAir       0
GarageCars       0
GarageArea       0
YrSold           0
SaleType         0
SaleCondition    0
SalePrice        0
dtype: int64

In [10]:
data.columns

Index(['LotFrontage', 'LotArea', 'YearBuilt', 'TotalBsmtSF', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'Street', 'CentralAir', 'GarageCars',
       'GarageArea', 'YrSold', 'SaleType', 'SaleCondition', 'SalePrice'],
      dtype='object')

In [11]:
matrix = data.corr()
matrix.style.background_gradient (cmap = 'BrBG')

In [12]:
import statsmodels as sm

In [13]:
import sklearn as sl

In [14]:
from sklearn.linear_model import LinearRegression

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
data.columns

Index(['LotFrontage', 'LotArea', 'YearBuilt', 'TotalBsmtSF', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'Street', 'CentralAir', 'GarageCars',
       'GarageArea', 'YrSold', 'SaleType', 'SaleCondition', 'SalePrice'],
      dtype='object')

In [18]:
Y=data['SalePrice']

In [19]:
X=data[['LotFrontage', 'LotArea', 'YearBuilt', 'TotalBsmtSF', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageCars',
       'GarageArea', 'YrSold']]

In [20]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [21]:
X.columns

Index(['LotFrontage', 'LotArea', 'YearBuilt', 'TotalBsmtSF', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageCars', 'GarageArea', 'YrSold'],
      dtype='object')

This code uses the statsmodels library to calculate the Variance Inflation Factor (VIF) for a set of predictor variables. It first selects the predictor variables, adds a constant term, and then computes the VIF values for each variable, helping to assess multicollinearity in a regression analysis.

In [73]:
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Select the predictor variables for which you want to calculate VIF
X = data[['LotFrontage', 'LotArea', 'YearBuilt', 'TotalBsmtSF', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageCars',
       'GarageArea', 'YrSold']]

# Add a constant term to the predictor variables
X = sm.add_constant(X)

# Calculate the VIF for each predictor variable
vif = pd.Series([variance_inflation_factor(X.values, i) for i in range(X.shape[1])], index=X.columns)

# Print the results
print(vif)


const           2.298904e+06
LotFrontage     1.480013e+00
LotArea         1.280666e+00
YearBuilt       1.591249e+00
TotalBsmtSF     1.653478e+00
BedroomAbvGr    1.963830e+00
KitchenAbvGr    1.147056e+00
TotRmsAbvGrd    2.460465e+00
GarageCars      5.383746e+00
GarageArea      5.304722e+00
YrSold          1.007453e+00
dtype: float64


In summary, this VIF analysis suggests that there may be severe multicollinearity present in the dataset, particularly affecting the constant term and, to a lesser extent, some of the predictor variables like "GarageCars" and "GarageArea." Addressing multicollinearity may be necessary to ensure the reliability of regression analysis results.

In [72]:
data.head()

LotFrontage  LotArea  YearBuilt  TotalBsmtSF  BedroomAbvGr  KitchenAbvGr  \
0         65.0     8450       2003          856             3             1   
1         80.0     9600       1976         1262             3             1   
2         68.0    11250       2001          920             3             1   
3         60.0     9550       1915          756             3             1   
4         84.0    14260       2000         1145             4             1   

   TotRmsAbvGrd Street CentralAir  GarageCars  GarageArea  YrSold SaleType  \
0             8   Pave          Y           2         548    2008       WD   
1             6   Pave          Y           2         460    2007       WD   
2             6   Pave          Y           2         608    2008       WD   
3             7   Pave          Y           3         642    2006       WD   
4             9   Pave          Y           3         836    2008       WD   

  SaleCondition  SalePrice  
0        Normal     208500  
1        Normal     181500  
2        Normal     223500  
3       Abnorml     140000  
4        Normal     250000

In [91]:
X=data[['LotFrontage', 'LotArea', 'YearBuilt', 'TotalBsmtSF', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageCars', 'GarageArea', 'YrSold']]

In [97]:
import statsmodels.api as sm
regression_model = sm.OLS(data['SalePrice'], sm.add_constant(data['LotArea'])).fit()
regression_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.097
Model:                            OLS   Adj. R-squared:                  0.096
Method:                 Least Squares   F-statistic:                     128.8
Date:                Fri, 17 Mar 2023   Prob (F-statistic):           2.03e-28
Time:                        03:17:59   Log-Likelihood:                -15251.
No. Observations:                1201   AIC:                         3.051e+04
Df Residuals:                    1199   BIC:                         3.052e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.482e+05   3673.232     40.334      0.000    1.41e+05    1.55e+05
LotArea        3.2771      0.289     11.348      0.000       2.710       3.844
==============================================================================
Omnibus:                      449.464   Durbin-Watson:                   1.986
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2830.023
Skew:                           1.594   Prob(JB):                         0.00
Kurtosis:                       9.811   Cond. No.                     2.04e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.04e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [99]:
import statsmodels.api as sm
regression_model = sm.OLS(data['SalePrice'], sm.add_constant(data['TotalBsmtSF'])).fit()
regression_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.393
Model:                            OLS   Adj. R-squared:                  0.393
Method:                 Least Squares   F-statistic:                     776.6
Date:                Fri, 17 Mar 2023   Prob (F-statistic):          3.49e-132
Time:                        03:18:43   Log-Likelihood:                -15013.
No. Observations:                1201   AIC:                         3.003e+04
Df Residuals:                    1199   BIC:                         3.004e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        5.722e+04   4813.736     11.887      0.000    4.78e+04    6.67e+04
TotalBsmtSF   116.6239      4.185     27.868      0.000     108.413     124.834
==============================================================================
Omnibus:                      302.632   Durbin-Watson:                   1.970
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5893.728
Skew:                           0.650   Prob(JB):                         0.00
Kurtosis:                      13.774   Cond. No.                     2.95e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.95e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [100]:
import statsmodels.api as sm
regression_model = sm.OLS(data['SalePrice'], sm.add_constant(data['YrSold'])).fit()
regression_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                    0.5091
Date:                Fri, 17 Mar 2023   Prob (F-statistic):              0.476
Time:                        03:19:18   Log-Likelihood:                -15312.
No. Observations:                1201   AIC:                         3.063e+04
Df Residuals:                    1199   BIC:                         3.064e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.773e+06   3.63e+06      0.763      0.445   -4.36e+06     9.9e+06
YrSold     -1291.1856   1809.670     -0.713      0.476   -4841.658    2259.287
==============================================================================
Omnibus:                      503.421   Durbin-Watson:                   2.008
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2661.883
Skew:                           1.896   Prob(JB):                         0.00
Kurtosis:                       9.230   Cond. No.                     3.03e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.03e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [102]:
import statsmodels.api as sm
regression_model = sm.OLS(data['SalePrice'], sm.add_constant(data[['GarageCars', 'GarageArea']])).fit()
regression_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.434
Model:                            OLS   Adj. R-squared:                  0.433
Method:                 Least Squares   F-statistic:                     459.1
Date:                Fri, 17 Mar 2023   Prob (F-statistic):          9.83e-149
Time:                        03:20:18   Log-Likelihood:                -14971.
No. Observations:                1201   AIC:                         2.995e+04
Df Residuals:                    1198   BIC:                         2.996e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5.573e+04   4516.894     12.339      0.000    4.69e+04    6.46e+04
GarageCars  4.385e+04   5112.585      8.576      0.000    3.38e+04    5.39e+04
GarageArea   100.9115     17.984      5.611      0.000      65.628     136.195
==============================================================================
Omnibus:                      489.899   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4159.240
Skew:                           1.658   Prob(JB):                         0.00
Kurtosis:                      11.492   Cond. No.                     1.61e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.61e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [103]:
import statsmodels.api as sm
regression_model = sm.OLS(data['SalePrice'], sm.add_constant(data[['LotFrontage', 'LotArea', 'YearBuilt', 'TotalBsmtSF', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'GarageCars', 'GarageArea', 'YrSold']])).fit()
regression_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.700
Model:                            OLS   Adj. R-squared:                  0.698
Method:                 Least Squares   F-statistic:                     278.3
Date:                Fri, 17 Mar 2023   Prob (F-statistic):          4.06e-303
Time:                        03:21:12   Log-Likelihood:                -14589.
No. Observations:                1201   AIC:                         2.920e+04
Df Residuals:                    1190   BIC:                         2.926e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -2.266e+06   2.01e+06     -1.130      0.259    -6.2e+06    1.67e+06
LotFrontage    -26.3069     66.279     -0.397      0.692    -156.343     103.729
LotArea          0.8523      0.189      4.511      0.000       0.482       1.223
YearBuilt      557.1727     52.565     10.600      0.000     454.043     660.303
TotalBsmtSF     48.0605      3.795     12.665      0.000      40.615      55.506
BedroomAbvGr -1.503e+04   2288.558     -6.567      0.000   -1.95e+04   -1.05e+04
KitchenAbvGr -5.712e+04   6097.772     -9.368      0.000   -6.91e+04   -4.52e+04
TotRmsAbvGrd  2.401e+04   1273.501     18.855      0.000    2.15e+04    2.65e+04
GarageCars    2.056e+04   3945.882      5.210      0.000    1.28e+04    2.83e+04
GarageArea      16.2292     13.778      1.178      0.239     -10.802      43.260
YrSold         594.3873    998.104      0.596      0.552   -1363.852    2552.626
==============================================================================
Omnibus:                      371.301   Durbin-Watson:                   1.934
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            21272.832
Skew:                           0.574   Prob(JB):                         0.00
Kurtosis:                      23.586   Cond. No.                     1.96e+07
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.96e+07. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Dependent Variable (SalePrice): This is the variable you are trying to predict or explain.

Model Summary: This section provides an overview of the regression model's performance and fit.

R-squared (R-squared): The R-squared value measures the proportion of the variance in the dependent variable (SalePrice) that is explained by the independent variables (predictors) in the model. In this case, the R-squared value is not shown, but it would indicate how well the model fits the data. A higher R-squared value (close to 1) suggests a better fit.

Adjusted R-squared (Adj. R-squared): This is a modified version of R-squared that accounts for the number of predictors in the model. It penalizes the inclusion of irrelevant predictors. A higher adjusted R-squared value indicates a better fit while considering model complexity.

Coefficients Table: This table provides information about the coefficients of each predictor variable in the model.

const (Intercept): This is the intercept or constant term in the regression equation. It represents the estimated value of SalePrice when all predictor variables are zero.

Coefficients for Predictor Variables: Each predictor variable (e.g., LotFrontage, LotArea, YearBuilt, etc.) has a coefficient associated with it. These coefficients represent the estimated change in SalePrice for a one-unit change in the respective predictor, while holding all other predictors constant.

std err: Standard error is a measure of the precision of the coefficient estimates. Smaller standard errors indicate more precise estimates.

t and P>|t|: These columns provide the t-statistic and associated p-value for each coefficient. The t-statistic measures the number of standard errors the coefficient is away from zero. The p-value tests whether the coefficient is statistically different from zero. Lower p-values (typically < 0.05) suggest that the predictor is statistically significant.

[0.025 0.975]: These columns provide the 95% confidence interval for each coefficient. It gives a range of values within which the true population coefficient is likely to fall.

Model Fit Information: This section contains additional statistical measures related to the model fit and error.

Omnibus and Prob(Omnibus): These statistics test the normality of residuals (model errors). A low Prob(Omnibus) value suggests that the residuals are not normally distributed.

Skew and Kurtosis: These statistics measure the skewness and kurtosis of residuals. Normally distributed residuals have skewness close to zero and kurtosis close to 3.

Durbin-Watson: This statistic tests for autocorrelation in residuals. Values around 2 suggest no significant autocorrelation.

AIC and BIC: These are information criteria that help in model selection. Lower values indicate a better fit, considering model complexity.

Warnings and Notes: This section may contain warnings or additional information about the regression model, such as multicollinearity or other issues.

Overall, the regression analysis aims to explain how changes in the predictor variables (e.g., LotFrontage, LotArea, YearBuilt, etc.) are associated with changes in the dependent variable (SalePrice). The coefficients, their significance, and other statistics help us understand the relationships between these variables and assess the overall quality of the model's fit to the data.